File: src/leetCode/java/Solution323.java
package leetCode.java;

import java.util.Arrays;

//Union-Find O(VlogE) O(V), V = n, E = edges.length
class Solution323 {
    public int countComponents(int n, int[][] edges) {
        int[] parents = new int[n];
        Arrays.fill(parents, -1);
        for(int[] edge : edges){
            int a = find(edge[0], parents);
            int b = find(edge[1], parents);

            if(a != b){
                n--;
                parents[a] = b;
            }
        }
        return n;
    }

    public int find(int node, int[] parents){
        if(parents[node] == -1) return node;
        return find(parents[node], parents);
    }
}

File: src/leetCode/c++/Solution323.cpp
//Union-Find O(ElogE) O(V), V = n, E = edges.size()
class Solution323 {
public:
    int countComponents(int n, vector<pair<int, int>>& edges) {
        vector<int> parent(n, -1);
        int res = n;//union的數量, 接合時-1
        for(auto edge : edges){
            int a = find(edge.first, parent);
            int b = find(edge.second, parent);

            if(a != b) --res;

            parent[a] = b;
        }
        return res;
    }

    int find(int node, vector<int> parent){
        if(parent[node] == -1) return node;
        return parent[node] = find(parent[node], parent);
    }
};